### ISSUE TITLE
Style cop idea: prefer `map` over `each` + `<<`

### ISSUE THREAD

**Issue by vlad-pisanov** [rubocop/rubocop#11878]
I often encounter this pattern, especially amongst junior devs; they start with the `[]` literal, and then build up a list by repeatedly pushing elements into it. It's messy and inefficient.

```ruby
# bad
result = []
arr.each { |x| result << x + 42 }

# good
result = arr.map { |x| x + 42 }
```

**Comment by natematykiewicz** [comment]
Do you have a way to detect that multiple loops aren't pushing to a single array? Because I sometimes do this to cut down on array allocations. This cop would force me to use `map` instead of `each`, and then I'd have to concatenate the arrays together still, which results in N more array allocations and also N more object iterations (`+` has to iterate the `result` list again during concatenation).

```ruby
result = []
arr1.each { |x| result << x if something }
arr2.each { |x| result << x if something }
```

**Comment by ymap** [comment]
I've had some hesitations, but I decided to go ahead and create the PR. The following points were particularly complex. I would really appreciate any feedback.

- Given that `Enumerable#map` returns an `Array`, I limited the cop's reporting to cases with a statically determinable `Array` as the push destination.

- As previously noted, due to efficiency concerns, I restricted the cop's reporting to cases where the `Array` being pushed to was clearly initialized as empty. Of course, using `+=` for non-empty `Array` was considered, but I feel it's more appropriate as a specific option in the cop, if needed.

- For cases where `each`'s return value might be used, I skipped autocorrection. Although rare, prioritizing safety over simplicity seemed crucial.
